CIPs/0022.md

```diff
@@ -0,0 +1,119 @@
+
+The currently deployed version of the encoding can be improved:
+1. It only works correctly when the validator set is at its maximum size. 
+2. As it's designed to be SNARK-friendly, the encoding uses the Bowe-Hopwood SNARK-friendly hash function on the entire elected validator set. As a result, showing a double-signing proof currently requires hashing the entire elected validator set with a non EVM-friendly hash, having too high gas costs.
+3. It's vulnerable to a "future subcommittee" attack, where an attacker that compromises enough validators can pre-sign encodings with future epoch numbers and use them when their compromised subcommittee is elected in the future.
+
+This CIP upgrades the encoding to address all three topics.
+
+## Motivation
+
+- Support validator sets which are not full up to a specified maximum.
+- Create the necesasry infrastructure for on-chain slashing in the case of double-signing.
+
+## Specification
+
+We define a few helper functions:
+* `encode_integer_X_le` - encodes an unsigned_integer having X bits to bits in a little-endian order.
+* `encode_public_key_le` - encodes a BLS12-377 point to a compressed form in bits - the little-endian bits of the x coordinate and a single bit whether y is the greatest of the two possibilities.
+* `bits_be_to_bytes_le` - converts big-endian bits to little-endian bytes.
+* `bytes_le_to_bits_be` - converts little-endian bytes to big-endian bits.
+* `bytes_le_to_bits_le` - converts little-endian bytes to little-endian bits.
+* `BHHash` - Bowe-Hopwood hash, as specified in the [Zcash protocol specification](https://github.com/zcash/zips/blob/master/protocol/protocol.pdf), on the Edwards curve defined over the same base field as BLS12-377, as specified in the [Zexe paper](https://eprint.iacr.org/2018/962.pdf).
+* `Blake2Xs` - the Blake2s-based variant of Blake2X, defined in the [Blake2X paper](https://blake2.net/blake2x.pdf).
+* `pad(array, n, padding_element)` - enlarge the array to size `n`, adding `padding_element` in the new places.
+* `truncate(array, n)` - truncate the array to the first `n` elements.
+
+The encoding is currently implemented as:
+
+```
+Blake2Xs(
+    BHHash(
+        encode_integer_16_le(epoch number) || 
+        encode_integer_32_le(maximum_non_signers) || 
+        validator set
+    )
+)
+```
+
+where `validator_set` is the concatenation of the results of `encode_public_key_le` on each validator public key.
+
+We propose to change it to the following. We first define the parameters `MAX_VALIDATORS = 150` and `SEC_PARAM = 128`. We then specify the new encoding.
+
+```
+Blake2Xs(
+    encode_integer_16_le(epoch number) || 
+    BHHash(
+        bytes_le_to_bits_le(truncate(parent_hash), SEC_PARAM/8) ||
```

Comment:
The closing parenthesis for truncate is misplaced, this line should instead say:
```
        bytes_le_to_bits_le(truncate(parent_hash, SEC_PARAM/8)) ||
```

Reply:
Oops, thanks!